tianshou/data/buffer/manager.py

```diff
@@ -141,22 +142,24 @@ def add(
         # get index
         if buffer_ids is None:
             buffer_ids = np.arange(self.buffer_num)
-        # TODO: what is ptrs?
-        ep_last_idxs, ep_lens, ep_rews, ep_start_idxs = [], [], [], []
+
+        ep_add_at_idxs, ep_lens, ep_rews, ep_start_idxs = [], [], [], []
         for batch_idx, buffer_id in enumerate(buffer_ids):
-            ep_last_idx, ep_rew, ep_len, ep_start_idx = self.buffers[buffer_id]._add_index(
+            ep_add_at_idx, ep_rew, ep_len, ep_start_idx = self.buffers[
+                buffer_id
+            ]._update_buffer_state_after_adding_batch(
                 batch.rew[batch_idx],
                 batch.done[batch_idx],
             )
-            ep_last_idxs.append(ep_last_idx + self._offset[buffer_id])
+            ep_add_at_idxs.append(ep_add_at_idx + self._offset[buffer_id])
             ep_lens.append(ep_len)
             ep_rews.append(ep_rew)
             ep_start_idxs.append(ep_start_idx + self._offset[buffer_id])
-            self.last_index[buffer_id] = ep_last_idx + self._offset[buffer_id]
+            self.last_index[buffer_id] = ep_add_at_idx + self._offset[buffer_id]
             self._lengths[buffer_id] = len(self.buffers[buffer_id])
-        ep_last_idxs = np.array(ep_last_idxs)
+        ep_add_at_idxs = np.array(ep_add_at_idxs)
```

Comment:
what is the semantics of this variable? The docstring calls it the current index. If it's not the ep_last_idxs, what does it mean?

Reply:
The index at which the transition is added to the buffer. 
As there is ep_start_idx that indicates the first transition of the current episode, ep_last_idx should be the  index of the last transition in the episode. Whenever the current transition does not contain done, this is not the last index of the episode (as it continues) but the index at which to add the current transition.

Reply:
Marking for discussion in pair programming